Implement MKLGenerator

michalowski-arm opened the pull request:
This PR aims to fix the issue from #132395 by implementing a new `MKLGeneratorImpl` that stores a consistent, global `vslStream` for use in random numbers generation. This path was previously disabled due to a problem of repeating variates, caused by repeated reseeding of the MKL generator with variates from the `CPUGenerator`. This new implementation only seeds the `MKLGenerator` once using the `CPUGenerator`, and then keeps reusing the same `vslStream`, providing the full period of the RNG.

For the sake of reproducibility, the saving and restoring of the `MKLGenerator` has been linked to `CPUGenerator` state changes, and the former does not provide its own `get_state()` and `set_state()` functionality. The point was to keep the user experience identical to before -- they do not need to handle a separate `MKLGenerator` explicitly.

There already exists a test to check for repetition based on the script from #132395. It can be found `test_distribution.py` as `test_multinomial_sequential_draw()`. For the old (reseeded) implementation of the MKL `vslStream`, this test showed 21 repetitions. For this new implementation, the test gives 0 repetitions as expected.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @azahed98 @chenyang78

michalowski-arm committed:
Implement MKLGeneratorImpl

  * Implements MKLGeneratorImpl which uses MKL/OpenRNG to generate random
    variates and keeps a consistent global state.
  * Links MKLGenerator state change to CPUGenerator state changes

michalowski-arm committed:
Use MKLGeneratorImpl in DistributionKernels.cpp

michalowski-arm commented:
CC @nikhil-arm @fadara01 @agrawal-aka @malfet @CaoE

michalowski-arm commented:
@pytorchbot label "topic: not user facing"

nikhil-arm commented:
@pytorchbot label "ciflow/linux-aarch64"

nikhil-arm commented:
@pytorchbot label "ciflow/linux-aarch64"

CaoE commented on aten/src/ATen/native/cpu/DistributionKernels.cpp:
```diff
@@ -158,19 +158,32 @@ void exponential_kernel(TensorIteratorBase &iter, double lambda, std::optional<G
+          {
+            // See Note [Acquire lock when using random generators]
+            std::lock_guard<std::mutex> lock(mklGenerator->mutex_);
+            mklGenerator->get_stream_copy(stream);
```
Isn't get_stream_copy thread-safe ?

malfet commented:
Those failures can not be a coincidence, are they?

leslie-fang-intel commented:
Hi @michalowski-arm @CaoE, the CI failure should be related. Could you take a look?

michalowski-arm commented:
At least some of the failures are due to not handling `torch.manual_seed` appropriately. It resets CPUGenerator to the beginning of the input seed's sequence (and the tests are expecting that reset) but MKLGenerator is unaffected and just continues its old sequence.

michalowski-arm committed:
Link MKLGenerator seed change to CPUGenerator seed changes

michalowski-arm commented:
@pytorchbot label "ciflow/linux-aarch64"

CaoE commented on aten/src/ATen/mklrng/MKLGeneratorImpl.cpp:
```diff
@@ -0,0 +1,155 @@
+  : c10::GeneratorImpl{Device(DeviceType::CPU), DispatchKeySet(c10::DispatchKey::CPU)},
+    seed_(seed_in),
+    offset_(0) {
+    vslNewStream(&stream_, VSL_BRNG_MCG59, seed_);
```
Is the period of `VSL_BRNG_MCG59` long enough? Can we consider `VSL_BRNG_PHILOX4X32X10`, `VSL_BRNG_MT19937` ... ?

michalowski-arm replied:
Sure, do you have a preference between the two?

CaoE commented:
There are some errors in `test_distributions.py`, for example, `test_cdf_icdf_inverse`. My local tests show that using `VSL_BRNG_MT19937 ` can pass the test. There may be other tests that will fail and can be made to pass by slightly modifying sample length.

michalowski-arm committed:
Use VSL_BRNG_PHILOX4X32X10 instead of VSL_BRNG_MCG59

michalowski-arm committed:
Allow more samples for convergence in tests of exponential

fadara01 commented:
@CaoE @leslie-fang-intel 
Would you guys be able to advise about the clang-specific failures (at test time):
```
  × python setup.py egg_info did not run successfully.
  │ exit code: 2
  ╰─> [1 lines of output]
      Intel MKL function load error: __vslNewStreamEx.
      [end of output]
```

fadara01 reviewed (commented):
For the bazel failure, I suspect you need to add this to [BUILD.bazel](https://github.com/pytorch/pytorch/blob/main/BUILD.bazel)
```
filegroup(
    name = "aten_mklrng_cpp",
    srcs = glob([
        "aten/src/ATen/mklrng/*.cpp",
    ]),
)
```
similar to `mkl` [here](https://github.com/pytorch/pytorch/blob/main/BUILD.bazel#L185C1-L191C2)
and then add `aten_mklrng_cpp` [here](https://github.com/pytorch/pytorch/blob/main/BUILD.bazel#L354)

michalowski-arm commented:
@pytorchbot label "skip-url-lint"

fadara01 commented:
> Those failures can not be a coincidence, are they?

Hi @malfet! We have now fixed most of the previous failures - the only ones left are the `clang` ones, which we'll hopefully fix soon with some help from @CaoE / @leslie-fang-intel.
Would it possible for us to get some feedback about the approach taken here to fix #132395?

fadara01 commented:
cc: @knjmooney

fadara01 commented:
@pytorchbot rebase

fadara01 commented:
@michalowski-arm can we try reproducing this locally on an x86 machine?
Maybe we can get a longer stack trace or more insights about the failure.

You can pull the docker image from `pytorch/pytorch-linux-jammy-py3-clang15-asan` as done [here](https://github.com/pytorch/pytorch/actions/runs/14753660853/job/41418707835#:~:text=90e2f0da05deb1b09040f7c1ee78ad0f34911877%3A%20Pulling%20from-,pytorch/pytorch%2Dlinux%2Djammy%2Dpy3%2Dclang15%2Dasan,-55) then download the build and test artifacts for `linux-jammy-clang15-asan` from https://hud.pytorch.org/pr/151218
and then just follow the "Test" steps in the pipeline.

michalowski-arm commented:
> @michalowski-arm can we try reproducing this locally on an x86 machine? It might give us a longer stack trace or more insights about the failure.
> 
> You can pull the docker image from `pytorch/pytorch-linux-jammy-py3-clang15-asan` as done [here](https://github.com/pytorch/pytorch/actions/runs/14753660853/job/41418707835#:~:text=90e2f0da05deb1b09040f7c1ee78ad0f34911877%3A%20Pulling%20from-,pytorch/pytorch%2Dlinux%2Djammy%2Dpy3%2Dclang15%2Dasan,-55) then download the build and test artifacts for `linux-jammy-clang15-asan` from https://hud.pytorch.org/pr/151218 and then just follow the "Test" steps in the pipeline.

Yes, I've been looking into it

CaoE commented:
@michalowski-arm I downloaded the image locally `docker pull ghcr.io/pytorch/ci-image:pytorch-linux-jammy-py3-clang15-asan-90e2f0da05deb1b09040f7c1ee78ad0f34911877`, and reproduced the `Intel MKL function load error: __vslNewStreamEx.` as done [here](https://github.com/pytorch/pytorch/actions/runs/14753660853/job/41418707835#:~:text=90e2f0da05deb1b09040f7c1ee78ad0f34911877%3A%20Pulling%20from-,pytorch/pytorch%2Dlinux%2Djammy%2Dpy3%2Dclang15%2Dasan,-55). 
I found that the MKL 2021 used in the conda environment `py_3.10` of this image is relatively old. I manually deleted the old mkl and pip install mkl-static mkl-include, and then found that the issue was gone.

abullabib commented:
@CaoE thank you for your insights!

>I found that the MKL 2021.4.0 used in the conda environment py_3.10

do you guys have a preference as to which version we should update to?

fadara01 commented:
@CaoE thank you for your insights!

>I found that the MKL 2021.4.0 used in the conda environment py_3.10

do you guys have a preference as to which version we should update to?

CaoE commented:
@fadara01 I tried MKL 2025.1, but not sure if 2024.2.0 can also fix the error. Let's see the result of https://github.com/pytorch/pytorch/pull/154198.

CaoE commented:
@malfet I have tired to upgrade MKL to 2024.2.0 in CI https://github.com/pytorch/pytorch/pull/154198, but there are many "Inconsistent configuration parameters" appears.  The new MKL has made some changes to DFTI. The MKL version in pytorch releases is inconsistent with the MKL used by CI, which causes these errors to not be triggered in CI. 
May I know, why the MKL used in CI is not upgraded to 2024 like in release? The torch FFT ops seem to need to be fixed for these "Inconsistent configuration parameters" issues.

CaoE commented:
I added the fix in this PR https://github.com/pytorch/pytorch/pull/154198.

nikhil-arm commented:
@pytorchbot rebase

michalowski-arm committed:
Update test_distributions.py

michalowski-arm committed:
Update test_variance

nikhil-arm commented:
CI is passing, review comments are addressed. 1 Ci failure seems to be unrelated.
Ready for merge

bigPYJ1151 commented:
Hi @malfet @michalowski-arm, I hope you're both doing well! This PR is crucial for our vLLM x86 CPU backend, as we're currently pinned to Torch 2.6 due to MKL RNG being disabled in Torch 2.7. Your help in moving this PR forward would be greatly appreciated. Thanks!

fadara01 reviewed (dismissed):
LGTM
@malfet could you please have a look at this? This is crucial for performance for both AArch64 and x86

aditew01 commented:
@pytorchbot rebase

malfet commented:
@fadara01 can you help me understand how this could be used on ARM? (MKL is x86 only library, isn't it?)

michalowski-arm committed:
Advance global MKLGenerator state before generation

  * Added a local `main_stream` copy of the global VSLStream
    and changed the kernel logic to immediately advance the global
    stream by the number of generated elements upon obtaining the copy.
    This fixes an issue where previously calling these kernels in
    multiple concurrent python threads could yield identical generated
    sequences (since the advance in the global stream happened only
    after the generation step was finished).

jgong5 reviewed (dismissed):
small nits, otherwise LGTM. Next time, suggest to avoid force-push to make incremental review easier.

fadara01 committed:
Fix indentation

michalowski-arm committed:
Move MKLGenerator from ATen/mklrng/ to ATen/mkl/

aditew01 commented:
@yangw-dev @atalman @malfet  can someone help with testing the latest push with meta internal to see if this is still an issue?

fadara01 commented:
@yangw-dev @atalman @malfet - today's the first birthday of this PR, we think the meta internal build should pass now with the latest changes, could you please verify?

malfet commented:
@pytorchbot merge

fadara01 commented:
@malfet - this needs another stamp :)

malfet reviewed (dismissed):
Sorry, forgot that reviews are stripped after revert

malfet commented:
@pytorchbot merge

jeanschmidt commented:
Build failure on ARM64: missing `AT_MKL_ENABLED()` guard

`MKLGeneratorImpl.cpp` unconditionally includes `<mkl.h>`, which doesn't exist on ARM64. This breaks all downstream ATen targets on that architecture.

CMake doesn't hit this because it excludes `mkl/*.cpp` when `AT_MKL_ENABLED` is off. But non-CMake build systems (Buck, Bazel) compile all files in the directory and rely on source-level guards — the same pattern `SparseBlas.cpp` already uses in the same directory.

### Fix

Wrap `MKLGeneratorImpl.cpp` contents with:

```cpp
#include <ATen/Config.h>

#if AT_MKL_ENABLED()

// ... existing file contents ...

#endif // AT_MKL_ENABLED()
```

The header doesn't need a guard — per `ATen/mkl/README.md`, callers are responsible for guarding includes, and they already do.

jeanschmidt commented:
@pytorchbot revert -m "Breaks builds on other build systems" -c ghfirst

fadara01 commented:
@michalowski-arm - would you be able to push a fix please?
the revert message mentions failures on "Buck" too.

michalowski-arm committed:
Guard MKLGeneratorImpl.cpp behind AT_MKL_ENABLED()

michalowski-arm commented:
@fadara01 @aditew01 @jeanschmidt done

aditew01 commented:
@malfet can you please test if Meta internal works with the latest changes? The fix for the issue reported has been pushed.

aditew01 commented on aten/src/ATen/mkl/MKLGeneratorImpl.h:
```diff
@@ -0,0 +1,46 @@
+struct TORCH_API MKLGeneratorImpl : public c10::GeneratorImpl {
+  // Constructors
+  MKLGeneratorImpl(uint64_t seed_in = default_rng_seed_val);
+  ~MKLGeneratorImpl() override = default;
```
the destructor does not call `vslDeleteStream` ? 
should it be something like this: 
```
MKLGeneratorImpl::~MKLGeneratorImpl() {
  if (!stream_)
    vslDeleteStream(&stream_);
}
```

aditew01 commented on aten/src/ATen/mkl/MKLGeneratorImpl.cpp:
```diff
@@ -0,0 +1,158 @@
+/**
+ * Private clone method implementation
+ * See Note [Acquire lock when using random generators]
+ */
+MKLGeneratorImpl* MKLGeneratorImpl::clone_impl() const {
+  auto gen = new MKLGeneratorImpl();
+  return gen;
+}
```
this implementation is not a clone, right? it returns a new instance of MKLGeneratorImpl class, but discards the `seed_` ? 
it should preserve that: with something like: 
```
auto gen = new MKLGeneratorImpl(seed_);
```

michalowski-arm committed:
Fix copy/delete behaviour in MKLGenerator
